Add EIP: Portal Wire Protocol extensible generic overlay over discv5 (#8629)
ethereum/EIPs: EIPS/eip-7718.md

---
eip: 7718
title: Portal Wire Protocol a framework for discv5
description: A generic framework over discv5 which allows new overlay protocols to be built without reimplementing DHT primitives
author: Kolby ML (@KolbyML)
discussions-to: https://ethereum-magicians.org/t/eip-7718-portal-wire-protocol-a-framework-for-discv5/20230

> status: Draft

Reviewer: ![-6147779594829283016_121.jpg](https://github.com/user-attachments/assets/47822c19-746a-4a2c-9a09-b10cc540703e)

Reviewer 2: Thx

---
eip: 7718
title: Portal Wire Protocol a framework for discv5
description: A generic framework over discv5 which allows new overlay protocols to be built without reimplementing DHT primitives
author: Kolby ML (@KolbyML)
discussions-to: https://ethereum-magicians.org/t/eip-7718-portal-wire-protocol-a-framework-for-discv5/20230
status: Draft
type: Standards Track
category: Networking
created: 2024-06-06
---

## Abstract

> Discv5 is an extensible protocol which allows building new protocols on top of it. We propose a generic framework for building new DHT networks on top of Discv5 referred to as the `Portal Wire Protocol`

Reviewer 3: ```suggestion
Discv5 is an extensible protocol which allows building new protocols on top of it. We propose a generic framework for building new distributed hash table (DHT) networks on top of Discv5 referred to as the `Portal Wire Protocol`
```

## Motivation

> Discv5 is implemented in many languages, has built in packet encryption, used by consensus clients, and allows building sub-protocols using it's `TalkRequest` message type.

Reviewer 3: ```suggestion
Discv5 is implemented in many languages, has built in packet encryption, used by consensus clients, and allows building sub-protocols using its `TalkRequest` message type.
```

With the ubiquity of Discv5 in Ethereum and it being implemented in many languages, building new protocols using Discv5 as a base is beneficial as it enables quicker implementation times, stability as Discv5 is already used in production, as well as inheriting improvements and optimizations discovered by Discv5.

When building new DHT sub-protocols on Discv5, new sub-protocols often share the same primitive message types. By defining the framework `Portal Wire Protocol`, new networks can be built without re-implementing the DHT primitives reducing the development time and inheriting the performance optimizations of the base `Portal Wire Protocol` implementation.

## Specification

> 

Reviewer 3: If you're using the uppercase keywords (SHOULD, MUST, etc.) you should include the paragraph from the template:

https://github.com/ethereum/EIPs/blob/2ba7edfbc305bf825ef17c459ee6c9d04add9bf4/eip-template.md?plain=1#L55

Author: I am linking to the `ethereum/portal-network-specs` now, so this language won't be directly in the EIP, if you still think this warning should be included in this document let me know

Author: I am linking to the `ethereum/portal-network-specs` now, so this language won't be directly in the EIP, if you still think this warning should be included in this document let me know

Reviewer 2: > I am linking to the `ethereum/portal-network-specs` now, so this language won't be directly in the EIP, if you still think this warning should be included in this document let me know

offers the content to maximum `n` of the newly discovered nodes.

The process above should quickly saturate the area of the DHT where the content is located and naturally terminate as more nodes become aware of the content.

#### POKE Mechanism

When a node in the network is doing a content lookup, it will practically perform a recursive find using the `FindContent` and `Content` messages.
During the course of this recursive find, it may encounter nodes along the search path which do not have the content but for which the `content-id` does fall within their announced radius. These are nodes that should be interested in storing this content unless their radius was recently changed.

If the node doing the lookup successfully retrieves the content from another node, it should send an `Offer` message for that content to those interested nodes. This mechanism is designed to help spread content to nodes that may not yet be aware of it.

## Rationale

Reducing the amount of duplicate code for primitive DHT functions, to enable cleaner and quicker development of sub-protocols over Discv5

> TBD

Reviewer 3: ```suggestion
TBD <!-- TODO -->
```

## Security Considerations

> Needs discussion.

Reviewer 3: ```suggestion
Needs discussion. <!-- TODO -->
```

---
eip: 7718
title: Portal Wire Protocol a framework for discv5
description: A generic framework over discv5 which allows new overlay protocols to be built without reimplementing DHT primitives
author: Kolby ML (@KolbyML)
discussions-to: https://ethereum-magicians.org/t/eip-7718-portal-wire-protocol-a-framework-for-discv5/20230
status: Draft
type: Standards Track
category: Networking
created: 2024-06-06
---

## Abstract

Discv5 is an extensible protocol which allows building new protocols on top of it. We propose a generic framework for building new DHT networks on top of Discv5 referred to as the `Portal Wire Protocol`

## Motivation

Discv5 is implemented in many languages, has built in packet encryption, used by consensus clients, and allows building sub-protocols using it's `TalkRequest` message type.

With the ubiquity of Discv5 in Ethereum and it being implemented in many languages, building new protocols using Discv5 as a base is beneficial as it enables quicker implementation times, stability as Discv5 is already used in production, as well as inheriting improvements and optimizations discovered by Discv5.

> Building new DHT sub-protocols on Discv5 all share the same primitive message types. By defining a framework `Portal Wire Protocol` new networks can be built without reimplementing the DHT primitives reducing the workload for new networks, well inheriting the performance optimizations of the base `Portal Wire Protocol` implementation for all overlay networks.

Reviewer 4: ```suggestion
When building new DHT sub-protocols on Discv5 these often share the same primitive message types. By defining the framework `Portal Wire Protocol`, new networks can be built without re-implementing the DHT primitives reducing the development time and inheriting the performance optimizations of the base `Portal Wire Protocol` implementation.
```

```

### Nodes and Node IDs

Nodes in the portal network are represented by their [EIP-778 Ethereum Node Record (ENR)](./eip-778.md) from the Discovery v5 network. A node's `node-id` is derived according to the node's identity scheme, which is specified in the node's ENR. A node's `node-id` represents its address in the DHT.  Node IDs are interchangeable between 32 byte identifiers and 256 bit integers.

### Request - Response Messages

The messages in the protocol are transmitted using the `TALKREQ` and `TALKRESP` messages from the base [Node Discovery Protocol](https://github.com/ethereum/devp2p/blob/56a498ee34ee0fb69ffd33dda026d632af4c4048/discv5/discv5-wire.md#talkreq-request-0x05).

All messages in the protocol have a request-response interaction:

- Request messages **MUST** be sent using a `TALKREQ` message.
- Response messages **MUST** be sent using the corresponding `TALKRESP` message.

All messages are encoded as an [SSZ Union](https://github.com/ethereum/consensus-specs/blob/04f5ec595d78c0e3e43794fb7644c18f2584770d/ssz/simple-serialize.md#union) type.

```
message = Union[ping, pong, find_nodes, nodes, find_content, content, offer, accept]

> serialized_message = serialize(message)

Reviewer 5: In some other places in our specs I think we have been using this format to make it clear that this is SSZ serialization.

```suggestion
serialized_message = SSZ.serialize(message)
```

```

The `serialized_message` is the payload passed to the `request` field of the `TALKREQ` message or the `reponse` field of the `TALKRESP` message.

The type values for the `Union` are the SSZ Containers specified per message type.

The transmission of `content` data that is too large to fit a single packet is done over [uTP](../assets/eip-7718/bep_0029-rst_post.pdf).

#### Aliases

For convenience we alias:

- `ByteList` to `List[uint8, max_length=2048]`

#### Message Types

##### Ping (0x00)

Request message to check if a node is reachable, communicate basic information about our node, and request basic information about the recipient node.

```
selector     = 0x00

> ping         = Container(enr_seq: uint64, custom_payload: ByteList)

Reviewer 4: We probably want to call this `ByteList[2048]` as the SSZ spec already defines `ByteList[N]`. 

(I think we also might want to lower it in case we want to be strict here in accordance to the discv5 packet limit + protocol overhead.)

```

- `enr_seq`: The node's current sequence number of their ENR record.
- `custom_payload`: Custom payload specified per the network.

##### Pong (0x01)

Response message to Ping(0x00)

```
selector     = 0x01

> pong         = Container(enr_seq: uint64, custom_payload: ByteList)

Reviewer 4: idem ^

```

- `enr_seq`: The node's current sequence number of their ENR record.
- `custom_payload`: Custom payload specified per the network.

##### Find Nodes (0x02)

Request message to get ENR records from the recipient's routing table at the given logarithmic distances. The distance of `0` indicates a request for the recipient's own ENR record.

```
selector     = 0x02
find_nodes   = Container(distances: List[uint16, max_length=256])
```

- `distances`: a list of distances for which the node is requesting ENR records for.
    - Each distance **MUST** be within the inclusive range `[0, 256]`
    - Each distance in the list **MUST** be unique.

##### Nodes (0x03)

Response message to FindNodes(0x02).

```
selector     = 0x03
nodes        = Container(total: uint8, enrs: List[ByteList, max_length=32])
```

- `total`: The total number of `Nodes` response messages being sent.
- `enrs`: List of byte strings, each of which is an RLP encoded ENR record.
    - Individual ENR records **MUST** correspond to one of the requested distances.
    - It is invalid to return multiple ENR records for the same `node_id`.
    - The ENR record of the requesting node **SHOULD** be filtered out of the list.

> > Note: If the number of ENR records cannot be encoded into a single message, then they should be sent back using multiple messages, with the `total` field representing the total number of messages that are being sent.

Reviewer 4: I think we need to remove this note as we don't have a solution currently for this multiple messages while remaining discv5 compliant
